seatunnel-engine/seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/dag/logical/LogicalDagOptimizer.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.seatunnel.engine.core.dag.actions.UnknownActionException;
+
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class LogicalDagOptimizer {
+
+    /**
+     * The action ID needs to be regenerated because of the source, sink, and chain.
+     */
+    private final IdGenerator idGenerator = new IdGenerator();
+
+    private final Map<Long, LogicalVertex> newLogicalVertexMap;
+
+    private final List<LogicalVertex> logicalJobVertices;
+
+    private final Set<LogicalEdge> oldEdges;
+
+    /**
+     * key: old vertex id, value: new vertex id;
+     */
+    private final Map<Long, Long> oldToNewIdMap = new HashMap<>();
+
+    private final LogicalDag dag;
+
+    public LogicalDagOptimizer(final Map<Long, LogicalVertex> logicalVertexMap,
+                               final Set<LogicalEdge> edges,
+                               JobConfig jobConfig) {
+        this.newLogicalVertexMap = new HashMap<>(logicalVertexMap.size());
+        this.logicalJobVertices = new ArrayList<>(logicalVertexMap.values());
+        this.oldEdges = edges;
+        this.dag = new LogicalDag(jobConfig, idGenerator);
+    }
+
+    public LogicalDag expandAndOptimize() {
+        getSourceVertices().forEach(sourceVertex -> {
+            List<LogicalVertex> vertices = new ArrayList<>();
+            vertices.add(sourceVertex);
+            for (int index = 0; index < vertices.size(); index++) {
+                LogicalVertex vertex = vertices.get(index);
+                createChainedVertex(vertex);
+                vertex.getOutputs().forEach(edge -> vertices.add(edge.getTargetVertex()));
+            }
+        });
+        connectVertices();
+        dag.getLogicalVertexMap().putAll(newLogicalVertexMap);
+        return dag;
+    }
+
+    private void connectVertices() {
+        oldEdges.forEach(oldEdge -> {
+            Long oldInputVertexId = oldEdge.getInputVertex().getVertexId();
+            Long oldTargetVertexId = oldEdge.getTargetVertex().getVertexId();
+            LogicalVertex inputVertex = newLogicalVertexMap.get(oldToNewIdMap.get(oldInputVertexId));
+            LogicalVertex targetVertex = newLogicalVertexMap.get(oldToNewIdMap.get(oldTargetVertexId));
+            LogicalEdge edge = new LogicalEdge(inputVertex, targetVertex);
+            inputVertex.addOutput(targetVertex);
+            targetVertex.addInput(inputVertex);
+            dag.addEdge(edge);
+        });
+    }
+
+    private List<LogicalVertex> getSourceVertices() {
+        List<LogicalVertex> vertices = new ArrayList<>();
+        for (LogicalVertex vertex : logicalJobVertices) {
+            if (vertex.getInputs().size() == 0) {
+                vertices.add(vertex);
+            }
+        }
+        return vertices;
+    }
+
+    private void createChainedVertex(LogicalVertex logicalVertex) {
```

Comment:
why move create transform chain from execution plan to logical paln?

Reply:
My opinion is that the ExecutionPlan phase should not convert logical vertices to logical vertices.